Initial API for G-formula

JulinaM opened the pull request:
**Purpose**: Scope the API for G-formula as a module within causallib

**High Level Overview**: 
Refactored `base_GFormula.py` & `GFormula.py` and moved to "**time_varying**" module.

Commit:
base skeleton created for G-Method.

Commit:
Refactored base_GFormula.py & GFormula.py and moved to "time_varying" module.

Commit:
added docs

Commit:
added few methods.

Commit:
inheriting IndividualOutcomeEstimator instead.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+import abc
+from typing import Optional, Any, OrderedDict
+import pandas as pd
+from ..estimation.base_estimator import IndividualOutcomeEstimator
```
Use consistent way of importing modules. e.g. if you are using absolute imports be consistent about it

JulinaM replied:
got it.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+import pandas as pd
+from ..estimation.base_estimator import IndividualOutcomeEstimator
+from sklearn.base import BaseEstimator as SKLearnBaseEstimator
+from .regression_curve_fitter import RegressionCurveFitter #TODO Referenced from Survival module. Assuming we need have similar wrapper class for fitting curve
```
Comment this out for now

JulinaM replied:
Done.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+
+
+#TODO
+class CovariateModels(OrderedDict):
```
What is the purpose of this class? If you are trying to define the type for covariate models it may not be doing what you want

JulinaM replied:
I wanted to coerce the type of Models for each Covariate variables. I'll just leave default OrderedDict for now.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+            outcome_model(IndividualOutcomeEstimator): A causal model that estimate on individuals level
+                                                      (e.g. Standardization).
+            treatment_model (???):
+            covariate_models (dict): {
```
`dict` -> `OrderedDict`

JulinaM replied:
Done

pchakraborty1 replied:
change the documentation as well

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+        self.covariate_models = covariate_models #TODO call _init_regressor() for all covariate_models
+        self.refit_models = refit_models
+
+    def _init_regressor(self, model):
```
Comment this part out --> we may not need this

JulinaM replied:
Okay. Using default init for now.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+    @abc.abstractmethod
+    def fit(self,
+            X: pd.DataFrame,
+            a: pd.Series,
```
@ehudkr G-computation can support multiple treatment variables. Do you think we can support a dataframe for `a`?

ehudkr replied:
Good point. I suggest we start with a single treatment at the moment, and expand it later?
(also, you mean concurrent treatment I guess, because otherwise a vector `a` can just take non-binary values, no?)

pchakraborty1 replied:
yes i was thinking of back to back or combination of treatments. e.g. vasopressor and fluid (both could be binary)
- we can certainly start with a single one and then expand

pchakraborty1 replied:
Comment added for single treatment variable

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+            a: pd.Series,
+            t: pd.Series,
+            y: Optional[Any] = None,
+            refit_models=True,
```
add the type hint for `bool`

JulinaM replied:
Added.

pchakraborty1 replied:
add the type hint to function call as well. currently only added to the documentation

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+    def estimate_individual_outcome(self,
+                                    X: pd.DataFrame,
+                                    a: pd.Series,
+                                    t: pd.Series,
```
`t` can be optional

pchakraborty1 replied:
done

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+                                    timeline_end: Optional[int] = None
+                                    ) -> pd.DataFrame:
+        """
+              Returns individual estimated curves for each subject row in X/a/t
```
@JulinaM  seems like the description for this and individual outcome is the same - please check

JulinaM replied:
Thanks. Changes made in the description.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+    model: IndividualOutcomeEstimator
+
+
+class GFormulaBase(IndividualOutcomeEstimator):
```
what functions are we deriving from `IndividualOutcomeEstimator`

JulinaM replied:
I'm trying to overload the functions from IndividualOutcomeEstimator as per discussion with Ehud. We need to confirm with @ehudkr about this.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+
+            self.treatment_model.fit(df=X, duration_col=t, event_col=y,
+                                    **kwargs)
+            self.treatment_model.fit(X, a)
```
These two statements doesn't look right - it should be either one of them, right?

JulinaM replied:
Agree. Thanks for letting me know. Just removed the redundant line.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,199 @@
+
+
+            for cov in self.covariate_models:
+                self.covariate_models[cov].fit(X, a)
```
For G-computation we may need to a helper to massage the data. the fit should be  on something like

`concat(X, x_1_lagged), a`  and so on. 

Perhaps we should add a helper function - possibly reuse the prepare data function

ehudkr replied:
yes. I believe the entire fitting process is probably more elaborate than that.
I took this PR as just for the basic structure rather than the actual content, no?

JulinaM replied:
@ehudkr Yes, this is just a basic structure. The more logic will be added here like preparing data to actually feed in the models.

Commit:
refactored code as per comments on jun 29

ehudkr commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,218 @@
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def apply_noise(self):
```
1. probably should not be in `Base` since it's a specific implementation of the monte carlo version of G-formula.
2. should probably be private method: `_apply_noise`

JulinaM replied:
I've made this method abstract in base class, so that we could have specific implementation for each sub classes.
I'm okay if you want me to implement as a private method.

pchakraborty1 replied:
yes lets move it to the specific implementation of GFormula

ehudkr commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,218 @@
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
```
That's a good point. I still don't have a good idea.
My current guess is that `a` defines a treatment regime. So we can return sample-id as index and time as column. And this result is per the inputted treatment.

pchakraborty1 replied:
that makes sense - we can clarify this in the documentation. Proposed doc change

```
if t is not provided, index from X is used as t. otherwise t is used
```

ehudkr commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,218 @@
+                                      If None, will predict up to last observed event (t.max()).
+
+              Returns:
+                  pd.DataFrame: with time-step index, subject IDs (X.index) as columns and ??
```
Following the logic above, (again, I'm partly speculating) here we might have a `Series`. time as index and counterfactual outcome (corresponding to the treatment regime at each time point) as values.

pchakraborty1 replied:
@JulinaM following above, add the same documentation for t and make it optional

ehudkr commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,218 @@
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None) -> pd.DataFrame:
+
+        min_time = timeline_start if timeline_start is not None else int(t.min())
+        max_time = timeline_end if timeline_end is not None else int(t.max())
+
+        contiguous_times = pd.Series(data=range(min_time, max_time + 1), name=t.name)  # contiguous time steps for inference
```
I think it might be possible to opt for a design in which the `time` is defined by the treatment assignment. Which on inference time is a desired treatment trajectory.

ehudkr reviewed (commented):
Looks like a good start laying out the basic structure.

Commit:
cleaned up the skeleton and docs.

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -143,7 +144,7 @@ def _predict_trajectory(self, X, a, t) -> pd.DataFrame:
 class GFormula(GFormulaBase):
     """
-        GFormula class that is based on Monte Carlo Simulation for creating noise. #TODO more on this
+        GFormula class that is based on Monte Carlo Simulation for creating the noise.
```
From @JulinaM : create follow up task for description about class

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,225 @@
+                                    y: Optional[Any] = None,
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None
+                                    ) -> pd.DataFrame:
```
Add a variable for treatment strategy. Possible solution:

```
treatment_strategy: Callable,
```

ehudkr replied:
I thought the strategy could be defined according to the provided `a`. Can we discuss it on our next call?

pchakraborty1 replied:
sure lets do that. `a` can capture the historical treatments but it needs a way to get the function to get the next one

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,225 @@
+                                    y: Optional[Any] = None,
+                                    timeline_start: Optional[int] = None,
+                                    timeline_end: Optional[int] = None
+                                    ) -> pd.DataFrame:
```
Like above, add a new variable for treatment_strategy

pchakraborty1 commented on causallib/time_varying/GFormula.py:
```diff
@@ -0,0 +1,225 @@
+            refit_models: bool = True,
+            **kwargs
+            ):
+
```
Add a `raise NotImplementedError()` for dummy codes

Commit:
1. Renamed base_GFormula.py to base.py
2.Renamed class name GFormula to BaseGMethod and moved it to base.py
3. Added treatment_strategy in the estimate_individual_outcome and estimate_population_outcome methods.
4. moved _apply_noise from abstract class to subclass

Commit:
Renamed class name BaseGFormula to GMethodBase

Commit:
Renamed class name BaseGFormula to GMethodBase

Commit:
Refactored docs for treatment desc in estimate methods.

Commit:
Moved "NotImplementedError" before the rough implementation

JulinaM commented:
@pchakraborty1 All the required changes has been made. Could you please review and accept the PR?

prithwi committed:
removing references to survival models

pchakraborty1 reviewed (approved):
Changes made according to initial discussions. API may evolve over time
